test/client/recordUtil.js

```diff
@@ -442,6 +442,24 @@ test('recordUtil.resolveRecords()', (t) => {
     t.deepEquals(resolved, [expectedRecord], t.name)
   })
 
+  t.test(`${t.name} Create + Update + Update of an existing object should resolve to a separate Update`, (t) => {
+    t.plan(1)
+
+    var createBookmark = recordBookmark
+    createBookmark.action = proto.actions.CREATE
+    var existingObject = recordBookmark
+
+    var updateBookmark1 = updateBookmark
+    updateBookmark1.bookmark.site.title = 'Title1'
+    var updateBookmark2 = updateBookmark
+    updateBookmark2.bookmark.site.title = 'Title2'
+
+    const input = [[createBookmark, existingObject], [updateBookmark1, existingObject], [updateBookmark2, existingObject]]
+    const resolved = recordUtil.resolveRecords(input)
+    const expected = [updateBookmark2]
```

Comment:
isn't this supposed to be `[createBookmark, updateBookmark2]`?

Reply:
expected should not have `createBookmark` because https://github.com/brave/sync/blob/merge_by_id_action_staging/client/recordUtil.js#L214 , when record has action CREATE and existingObject is not empty, it is ignored.

Reply:
About `sequential Updates should become no op` test, https://github.com/brave/sync/commit/4af5ad43572e5d80db7e565a9ff8196489b84c84#diff-dbbb41acf4ed22ffd4a6afc45647bb64R408 

We just do not touch this test execution, because there are no CREATE operation in record. Here is a log
```
# recordUtil.resolveRecords() sequential Updates should become no op
input=[
[
{"action":1,"deviceId":{"0":0},"objectId":{"0":142,"1":158,"2":77,"3":121,"4":46,"5":224,"6":77,"7":81,"8":17,"9":15,"10":146,"11":30,"12":139,"13":27,"14":182,"15":192},"objectData":"bookmark","bookmark":{"site":{"location":"https://www.jisho.org","title":"jisho","customTitle":"辞書","lastAccessedTime":1480004209001,"creationTime":1480004209001,"favicon":"foobar"}}},
{"action":0,"deviceId":{"0":0},"objectId":{"0":142,"1":158,"2":77,"3":121,"4":46,"5":224,"6":77,"7":81,"8":17,"9":15,"10":146,"11":30,"12":139,"13":27,"14":182,"15":192},"objectData":"bookmark","bookmark":{"site":{"location":"https://www.jisho.org","title":"jisho","customTitle":"a ball pit filled with plush coconuts","lastAccessedTime":1480004209001,"creationTime":1480004209001,"favicon":"foobar"}}}
],
[
{"action":1,"deviceId":{"0":0},"objectId":{"0":142,"1":158,"2":77,"3":121,"4":46,"5":224,"6":77,"7":81,"8":17,"9":15,"10":146,"11":30,"12":139,"13":27,"14":182,"15":192},"objectData":"bookmark","bookmark":{"site":{"customTitle":"a ball pit filled with plush coconuts"}}},
{"action":0,"deviceId":{"0":0},"objectId":{"0":142,"1":158,"2":77,"3":121,"4":46,"5":224,"6":77,"7":81,"8":17,"9":15,"10":146,"11":30,"12":139,"13":27,"14":182,"15":192},"objectData":"bookmark","bookmark":{"site":{"location":"https://www.jisho.org","title":"jisho","customTitle":"a ball pit filled with plush coconuts","lastAccessedTime":1480004209001,"creationTime":1480004209001,"favicon":"foobar"}}}]]
Ignoring UPDATE of object 142,158,77,121,46,224,77,81,17,15,146,30,139,27,182,192.
resolved=[]
ok 55 recordUtil.resolveRecords() sequential Updates should become no op
```

so (`update1` merge `update2`) becomes equal to `existingObject` , `"customTitle"` is taken from update2.

Reply:
About `Create + Update of a new object should resolve to a single Create` https://github.com/brave/sync/commit/4af5ad43572e5d80db7e565a9ff8196489b84c84#diff-dbbb41acf4ed22ffd4a6afc45647bb64R429 

In this test either typo in description or in code.
Here is a log:
```
# recordUtil.resolveRecords() Create + Update of a new object should resolve to a single Create

input=[
[
{"action":1,"deviceId":{"0":0},
"objectId":{"0":23,"1":13,"2":212,"3":91,"4":136,"5":47,"6":225,"7":129,"8":170,"9":146,"10":219,"11":58,"12":239,"13":106,"14":119,"15":39},"objectData":"bookmark","bookmark":{"isFolder":false,"hideInToolbar":true,"site":{"location":"https://www.jisho.org","title":"jisho","customTitle":"辞書","lastAccessedTime":1480004209001,"creationTime":1480004209001,"favicon":"foobar"}}},
null
],
[
{"action":1,"deviceId":{"0":0},
"objectId":{"0":23,"1":13,"2":212,"3":91,"4":136,"5":47,"6":225,"7":129,"8":170,"9":146,"10":219,"11":58,"12":239,"13":106,"14":119,"15":39},"objectData":"bookmark","bookmark":{"site":{"customTitle":"a ball pit filled with plush coconuts"}}},
null]
]

resolved=
[
{"action":0,"deviceId":[0],"objectId":[23,13,212,91,136,47,225,129,170,146,219,58,239,106,119,39],"objectData":"bookmark","bookmark":{"isFolder":false,"hideInToolbar":true,"site":{"title":"jisho","favicon":"foobar","customTitle":"a ball pit filled with plush coconuts","lastAccessedTime":1480004209001,"creationTime":1480004209001,"location":"https://www.jisho.org"}}}
]

ok 56 recordUtil.resolveRecords() Create + Update of a new object should resolve to a single Create
```

So in fact this is `*Update* + Update of a new object should resolve to a single Create` and my PR did not touched 

But thanks @darkdh you pointed this place. Updating the testcase code.

Reply:
After fixing test I see to be an actual [CREATE, UPDATE] , it fails.

Sync library does conversion UPDATE=>CREATE when sees resolve [UPDATE, null]
https://github.com/brave/sync/blob/merge_by_id_action_staging/client/recordUtil.js#L220 

But it can perform such convertion only if bookmark record has non-empty title for folders or non-empty location for regular bookmsrks
https://github.com/brave/sync/blob/merge_by_id_action_staging/client/recordUtil.js#L51

in the testcase UPDATE has no field of location, so this UPDATE is ignored. 

So my PR changes of have separate CREATE and UPDATE records on merge will be modified to keep CREATE and UPDATE as separate only if existing object is not null.

Reply:
Pushed two commits for this